proton/cert_pinning.py

```diff
@@ -0,0 +1,171 @@
+from OpenSSL import crypto
+from requests.adapters import HTTPAdapter
+from urllib3.connectionpool import HTTPSConnectionPool
+from urllib3.poolmanager import PoolManager
+from urllib3.util.timeout import Timeout
+
+from .constants import PUBKEY_HASH_DICT
+
+
+class TLSPinningError(requests.exceptions.SSLError):
+    def __init__(self, strerror):
+        self.strerror = strerror
+        super(TLSPinningError, self).__init__(strerror)
+
+
+class TLSPinningHTTPSConnectionPool(HTTPSConnectionPool):
+    """HTTPSConnectionPool that verifies the certificate for each connection"""
+    def __init__(
+        self,
+        host,
+        hash_dict,
+        port=None,
+        strict=False,
+        timeout=Timeout.DEFAULT_TIMEOUT,
+        maxsize=1,
+        block=False,
+        headers=None,
+        retries=None,
+        _proxy=None,
+        _proxy_headers=None,
+        key_file=None,
+        cert_file=None,
+        cert_reqs=None,
+        key_password=None,
+        ca_certs=None,
+        ssl_version=None,
+        assert_hostname=None,
+        assert_fingerprint=None,
+        ca_cert_dir=None,
+        **conn_kw
+    ):
+        self.hash_dict = hash_dict
+        super(TLSPinningHTTPSConnectionPool, self).__init__(
+            host,
+            port,
+            strict,
+            timeout,
+            maxsize,
+            block,
+            headers,
+            retries,
+            _proxy,
+            _proxy_headers,
+            key_file,
+            cert_file,
+            cert_reqs,
+            key_password,
+            ca_certs,
+            ssl_version,
+            assert_hostname,
+            assert_fingerprint,
+            ca_cert_dir,
+            **conn_kw
+        )
+
+    def _validate_conn(self, conn):
+        r = super(TLSPinningHTTPSConnectionPool, self)._validate_conn(conn)
+
+        sock = conn.sock
+
+        pem_certificate = self.get_certificate(sock)
+
+        if self.is_session_secure(pem_certificate, conn, self.hash_dict):
+            return r
+
+    def is_session_secure(self, cert, conn, hash_dict):
+        """Checks if connection is secure"""
+
+        cert_hash = self.extract_hash(cert)
+
+        if not self.validate_hash(cert_hash, hash_dict):
+            # Also generate a report
+            conn.close()
+            raise TLSPinningError("Insecure connection")
+
+        return True
+
+    def validate_hash(self, cert_hash, hash_dict):
+        """Validates the hash agains a known list of hashes/pins"""
+        try:
+            hash_dict[self.host].index(cert_hash)
+        except (ValueError, KeyError, TypeError):
+            return False
+        else:
+            return True
+
+    def get_certificate(self, sock):
+        """Extracts and converts certificate to PEM"""
+        certificate_binary_form = sock.getpeercert(True)
+        pem_certificate = DER_cert_to_PEM_cert(certificate_binary_form)
+
+        return pem_certificate
+
+    def extract_hash(self, cert):
```

Comment:
I personally like to visually split code blocks with newlines, but if there is no logical split, then there's no need for newline. Here, you're processing an object and it's pretty sequential. I'd remove the newlines :)

Reply:
Agreed :+1: